### ISSUE TITLE
`tailrec` to enforce tail recursion

### ISSUE THREAD

**Issue by WhiteBlackGoose** [LanguageDev/.NET-language-ideas#11]
Stole from Kotlin

**Comment by LPeter1997** [comment]
Agree, things like this are useful. One thing I'd like is to push these to metadata/attribute level and not have them occupy a keyword.

**Comment by aloisdg** [comment]
F# has a rec keyword. Why not re-use it?

**Comment by LPeter1997** [comment]
@aloisdg The `rec` keyword in F# is essentially heritage from the ML family and denotes that a binding is recursive (can reference itself, like a recursive function). I'm not a fan of the feature, but that's a topic for another discussion. This feature is about telling the compiler that TCO can - and should - be applied to a function to avoid an expensive recursive call and an exploding call stack. This is why I'm not a fan of making it a keyword, as it's more like communication with the compiler [that C# already does with attributes](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.methodimplattribute?view=net-6.0).

**Comment by WhiteBlackGoose** [comment]
Yep, and F# doesn't have this feature :(

**Comment by jl0pd** [comment]
Making it an attribute will require allowing attributes on local functions, because tailrecs are rarely used as top level function, mostly they're inner functions. Take F# [list module](https://github.com/dotnet/fsharp/blob/main/src/FSharp.Core/list.fs) as an example (search for "loop"). Attributes on local functions will look too heavyweight

**Comment by LPeter1997** [comment]
How so? Allowing attributes everywhere seem simply reasonable.

**Comment by 333fred** [comment]
One thing to make sure of: iirc the CLR does _not_ guarantee that it can always `tailrec`, even with the `.tail` directive. This lack of guarantee is one reason why C# does not have it.

**Comment by WhiteBlackGoose** [comment]
F# does it on IL level, and that's how I want to have it too.

**Comment by 333fred** [comment]
> F# does it on IL level, and that's how I want to have it too.

And because the clr doesn't guarantee `.tail` is always respected, it can sometimes fail.

**Comment by WhiteBlackGoose** [comment]
My point is that we don't need to emit IL's tail call at all. We can turn our recursion into a loop, basically, and only then compile it to IL

**Comment by 333fred** [comment]
> My point is that we don't need to emit IL's tail call at all. We can turn our recursion into a loop, basically, and only then compile it to IL

You can, but that (as far as I know) isn't how F# does it. It uses `.tail`.

**Comment by WhiteBlackGoose** [comment]
It emits tail very rarely. [Simple example](https://sharplab.io/#v2:DYLgZgzgNAJiDUAfAsAKGAUwC4AIBOGAxjgIYww7ACWEuJhxAvGjqzgLYlaEAWlNuAO5UsPFm0Q4A2gF0cAWgB8pBuNaSecEFmALlZCjpwAKesXg5NASjRA=) 

That's exactly why unlike C#, it can almost guarantee TCO for tailrec.

**Comment by jl0pd** [comment]
> How so? Allowing attributes everywhere seem simply reasonable.
> 
> ```swift
> func fact(n: int32): int32 {
>     #[TailRecursive]
>     func f(acc: int32, n: int32): int32 =
>         if (n == 0) acc
>         else f(n * acc, n - 1);
>     return f(1, n);
> }
> ```
> 
> Seems quite elegant IMO.

Less elegant than F# version

```fsharp
let fact n =
    let rec loop acc x =
        if x = 0 
        then acc
        else loop (acc * x) (x - 1)
    loop 1 n
```
